core/types/calls.cc

```diff
@@ -680,6 +692,129 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, DispatchArgs args,
+                                gs.beginError(core::Loc(args.locs.file, args.locs.call), errors::Infer::UntypedSplat)) {
+                            e.setHeader("Passing a hash where the specific keys are unknown to a method taking keyword "
+                                        "arguments");
+                            e.addErrorSection(ErrorSection("Got " + kwSplatType->show(gs) + " originating from:",
+                                                           kwSplatArg->origins2Explanations(gs)));
+                            result.main.errors.emplace_back(e.build());
+                        }
+                        kwargs = Types::untypedUntracked();
+                    }
+                }
+
+                // Check to see if the keyword splat was a valid kwargs hash, and consume a positional argument if it
+                // was implicit.
+                if (implicitKwsplat && kwargs != nullptr) {
+                    --posArgs;
+                }
+            } else {
+                // This function doesn't take keyword arguments, so consume the kwsplat and mark all the keyword args as
+                // an untyped hash.
+                kwargs = Types::hashOfUntyped();
+                --aend;
+            }
+        } else {
+            kwargs = make_type<ShapeType>(Types::hashOfUntyped(), move(keys), move(values));
+        }
+
+        // Detect the case where not all positional arguments were supplied, causing the keyword args to be consumed as
+        // a positional hash.
+        if (kwargs != nullptr && pit != pend && !pit->flags.isBlock) {
+            if (!hasKwargs || (!pit->flags.isRepeated && !pit->flags.isKeyword && !pit->flags.isDefault)) {
+                // TODO(trevor) if `hasKwargs` is true at this point but not keyword args were provided, we could add an
+                // autocorrect to turn this into `**kwargs`
+
+                // If there are positional arguments left to be filled, but there were keyword arguments present,
+                // consume the keyword args hash as though it was a positional arg.
+                if (auto e = matchArgType(gs, *constr, core::Loc(args.locs.file, args.locs.call),
+                                          core::Loc(args.locs.file, args.locs.receiver), symbol, method,
+                                          TypeAndOrigins{kwargs, {kwargsLoc}}, *pit, args.selfType, targs, kwargsLoc,
+                                          args.args.size() == 1)) {
+                    result.main.errors.emplace_back(std::move(e));
+                }
+
+                if (!pit->flags.isRepeated) {
+                    pit++;
+                }
+
+                // Clear out the kwargs hash so that no keyword argument processing is triggered below, and also mark
+                // the keyword args as consumed.
+                kwargs = nullptr;
+                ait += numKwargs;
+                posArgs++;
+            }
+        }
+    }
```

Comment:
This is really the meat of the change: it's simulating the old behavior by reconstructing a ShapeType from the keyword arguments.